rfd/001-tcp-load-balancer-design-doc.md

```diff
@@ -0,0 +1,119 @@
+
+### Request Forwarder
+
+Incoming packets will be forwarded after checking 3 criterion:
+* Client must be authenticated.
+* Client must be within connection limit.
+* Client must be authorized to access at least one healthy host.
+
+Once these requirements are satisfied, the request will be forwarded to an authorized host with the least connections.
+
+The least connection tracking will prioritize simplicity over performance for this project. Each new connection that is made will increment a host's connection count, and each completed connection will decrement the host's connection count. Each incoming request will iterate through all authorized hosts to find the one with the least active connections. A more performant solution could later be devised which may not have to iterate over every authorized host for every request (likely leveraging an async queue). One additional shortcoming of this simple approach is that a large batch of requests which arrive at exact same time could be routed to the exact same host, but this initial project is not planning to handle that edge case.
+
+Example of simple approach for selecting host:
+```
+func (l *LoadBalancer) LeastConnections() *upstream.TcpHost {
+	if l == nil || len(l.hosts) == 0 {
+		return nil
+	}
+
+	host := l.hosts[0]
+
+	for i, h := range l.hosts[1:len(hosts)] {
+		if h.ConnectionCount() < host.ConnectionCount() {
+			host = h
+		}
+	}
+
+	return host
+}
+```
+
+Once the host is selected, data will be passed directly from clients to upstream hosts, without being stored on the load balancer.
+
+### Health Checks
+The load balancer removes unhealthy upstreams if it is unable to connect to the host while handling a client connection. The health checks could include a grace period, or some request-forwarding backoff, allowing for n attempts before removal, but for simplicity, this project will immediately remove the host from available hosts upon a single failed connection.
+
+
+## Demonstrating Functionality
+
+The current plan (subject to change with implementation) to demonstrate successful functionality of the project will be to initialize and connect a few hosts and a client as part of the main.go file for the load balancer. This should provide a quick way to test local development while iterating on a solution. Log statements will output actions for simple observability. The app will be run with `go run main.go`, and a pre-built binary to work on 64-bit Linux machines will eventually be included once the app is complete. In a future version, the service would either allow configuration for upstream hosts, or expose methods for registering new upstream hosts, but that will be out of scope for now.
```

Comment:
I'd say we could leave off the pre-built binaries - README instructions for building are fine.

Reply:
Sounds good. I'll remove that and ensure we have some instructions in the README.